src/wireviz/wv_bom.py

```diff
@@ -78,101 +93,89 @@ def generate_bom(harness):
+    bom_entries.extend(harness.additional_bom_items)
 
     # remove line breaks if present and cleanup any resulting whitespace issues
     bom_entries = [{k: clean_whitespace(v) for k, v in entry.items()} for entry in bom_entries]
 
     # deduplicate bom
     bom = []
-    bom_types_group = lambda bt: (bt['item'], bt['unit'], bt['manufacturer'], bt['mpn'], bt['pn'])
-    for group in Counter([bom_types_group(v) for v in bom_entries]):
-        group_entries = [v for v in bom_entries if bom_types_group(v) == group]
-        designators = []
-        for group_entry in group_entries:
-            if group_entry.get('designators'):
-                if isinstance(group_entry['designators'], List):
-                    designators.extend(group_entry['designators'])
-                else:
-                    designators.append(group_entry['designators'])
-        designators = list(dict.fromkeys(designators))  # remove duplicates
-        designators.sort()
-        total_qty = sum(entry['qty'] for entry in group_entries)
-        bom.append({**group_entries[0], 'qty': round(total_qty, 3), 'designators': designators})
-
-    bom = sorted(bom, key=lambda k: k['item'])  # sort list of dicts by their values (https://stackoverflow.com/a/73050)
-
-    # add an incrementing id to each bom item
-    bom = [{**entry, 'id': index} for index, entry in enumerate(bom, 1)]
-    return bom
-
-def get_bom_index(harness, item, unit, manufacturer, mpn, pn):
+    for _, group in groupby(sorted(bom_entries, key=bom_types_group), key=bom_types_group):
+        group_entries = list(group)
+        designators = sum((make_list(entry.get('designators')) for entry in group_entries), [])
+        total_qty = sum(entry.get('qty', 1) for entry in group_entries)
+        bom.append({**group_entries[0], 'qty': round(total_qty, 3), 'designators': sorted(set(designators))})
+
+    # add an incrementing id to each bom entry
+    return [{**entry, 'id': index} for index, entry in enumerate(bom, 1)]
+
+def get_bom_index(bom: List[BOMEntry], part: AdditionalComponent) -> int:
+    """Return id of BOM entry or raise StopIteration if not found."""
     # Remove linebreaks and clean whitespace of values in search
-    target = tuple(clean_whitespace(v) for v in (item, unit, manufacturer, mpn, pn))
-    for entry in harness.bom():
-        if (entry['item'], entry['unit'], entry['manufacturer'], entry['mpn'], entry['pn']) == target:
-            return entry['id']
-    return None
-
-def bom_list(bom):
-    keys = ['id', 'item', 'qty', 'unit', 'designators'] # these BOM columns will always be included
-    for fieldname in ['pn', 'manufacturer', 'mpn']: # these optional BOM columns will only be included if at least one BOM item actually uses them
+    target = tuple(clean_whitespace(v) for v in bom_types_group({**asdict(part), 'description': part.description}))
+    return next(entry['id'] for entry in bom if bom_types_group(entry) == target)
```

Comment:
- Maybe we need to catch the exception and re-raise a more descriptive one?
- Alternatively,we can revert this line and use the old loop, but raise an exception instead of returning None?

Reply:
Sorry, I have been very hands-off regarding the BOM generation code. Under what circumstances would this exception be triggered? Can it happen with the current code (where this function is only called from `﻿get_additional_component_table()` or only if it happens to be called from somewhere else?

Reply:
Without this PR, None is returned if no entry found matching the search arguments. I suggest raising an exception instead of returning None to avoid such an internal error be undetected by the user. My comment here just suggest two different ways to make a more user friendly exception in such a case that normally never happen

Reply:
- I have implemented the latter of the two alternatives in https://github.com/formatc1702/WireViz/pull/197/commits/9142c21e4860911cc977a114b1199e5f2442b1e4 .
- I have also added a few commits to prepare for my suggestion in https://github.com/formatc1702/WireViz/pull/226#issuecomment-805448151.